llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td

```diff
@@ -1554,3 +1554,21 @@ let Predicates = [HasCheri, IsRV64, IsCapMode] in {
 defm : CheriLdPat<load, CLC_128>;
 defm : CheriStPat<store, CSC_128, GPCR>;
 } // Predicates = [HasCheri, IsRV64, IsCapMode]
+
+
+/// CHERI memory colouring extensions
+let Predicates = [HasCheri, IsRV64] in {
+def CGetVersion   : Cheri_r<0x13, "cgetversion">;
+def CSetVersion   : Cheri_rr<0x2, "csetversion">;
+let mayLoad = 1 in
+{
+def CLoadVersion  : Cheri_r<0x14, "cloadversion">;
+def CLoadVersions : Cheri_r<0x15, "cloadversions">;
+}
+let mayStore = 1, mayLoad = 0, hasSideEffects = 0 in
+def CStoreVersion : RVInstCheriTwoSrc<0x7e, 0x2, 0, OPC_CHERI, (outs),
+                                (ins GPR:$rs1, GPCR:$rs2),
+                                "cstoreversion", "$rs1, $rs2">;
+let mayLoad = 1, mayStore = 1 in
+def CAmoCDecVersion : Cheri_rr<0x3, "camocdecversion", GPR>;
```

Comment:
Is it important to indicate that `$rs1` is atomic?

```suggestion
def CAmoCDecVersion : Cheri_rr<0x3, "camocdecversion", GPR, GPR, GPCRMemAtomic>;
```

Reply:
I believe that type exists only to allow using `0(a0)` in the assembler despite the instruction not having an immediate operand.

Reply:
I've not seen that before, but in other back ends you never specify anything explicitly on operands to indicate that they are the memory addresses for atomics.

Reply:
Yes it's solely a parser thing. RISC-V's atomic operands are the only memory-accessing instructions that don't take an immediate, and have `(rs1)` as the canonical form but also accept `0(rs1)`, primarily to make it easy for generated assembly. Since our explicit loads and stores end up looking the same in terms of operands I reused the existing class (and its corresponding capability version), and did not rename it to avoid merge conflicts just for something cosmetic. In this case you do want the suggested change, but only so its assembly syntax matches the other AMOs, i.e. that `0(cs1)` is permitted in the assembly.

Reply:
This being the case I am tempted to (ab)use `GPCRMemAtomic` for the address operands to cloadversion and cstore version too so that they get brackets around the address. Thoughts?

Reply:
Yeah, that's what I would do, then it matches the explicit loads/stores. Maybe one day I'll rename it to something more general (and do so upstream too, even if it _currently_ makes less sense, but I think the hypervisor load/store-as-guest instructions are also immediate-less and will end up wanting the same syntax).